### PR TITLE
Sequential Monte Carlo - Approximate Bayesian Computation

### PR DESCRIPTION
**Pull request by agustinaarroyuelo**
Progress done so far in the SMC-ABC module. 

As you may know, this was my Google summer of code project last year, mentored by @aloctavodia  and @junpenglao . Here is a link to my final evaluation [blogpost](https://agustinaarroyuelo.github.io/jekyll/update/2018/08/08/final_evaluation.html). 

This step-method has a lot of issues still, mainly regarding acceptance rates. I have included in this PR a notebook with the Lotka Volterra equations example, so than you can test. 

I will be very glad to read your comments and suggestions.

**Interesting papers:**

[Fundamentals and Recent Developments in Approximate Bayesian Computation](https://academic.oup.com/sysbio/article/66/1/e66/2420817)

[Approximate Bayesian Computation](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002803)

### CODE REVIEW

**Comment by twiecki** [comment]
Very exciting, the NB seems to have quite a few exceptions, however.

**Comment by michaelosthege** [comment]
Hi @agustinaarroyuelo ,
Looks like a cool addition! I think some of the new components, like the `Simulator` can be useful for a variety of scenarios.

I'll go through and leave comments directly in the diff.

For the notebook I'd like to suggest that you
1. reduce the number of data points to bring in more uncertainty
2. plot the posterior predictive distribution (either as a credibility band, or as thin lines with `alpha=0.05` or so)

File: pymc3/distributions/simulator.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import numpy as np
+from .distribution import NoDistribution, Distribution
+
+__all__ = ['Simulator']
+
+
+class Simulator(NoDistribution):
+
+    def __init__(self, function, *args, **kwargs):
+
```
Comment (michaelosthege) [inline pymc3/distributions/simulator.py:19]: The `Simulator` class, and most importantly its `__init__` need a docstring.
What is the signature of the `function`
+ Can it take arguments?
+ Can it return any kind of data, or does it have to be a ?-dimensional array?

File: pymc3/distributions/simulator.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            Desired size of random sample (returns one sample if not
+            specified).
+
+        Returns
+        -------
+        array
+        """
+
+        print('Not implemented yet')
+
```
Comment (michaelosthege) [inline pymc3/distributions/simulator.py:50]: Fail fast: `raise NotImplementedError('Who is in charge of implementing this?')`

File: pymc3/sampling.py [hunk 3]
```diff
@@ -335,6 +335,15 @@ def sample(draws=500, step=None, init='auto', n_init=200000, start=None, trace=N
                                progressbar=progressbar,
                                model=model,
                                random_seed=random_seed)
+        
+    elif isinstance(step, pm.step_methods.smc_ABC.SMC_ABC):
+        if step_kwargs is None:
+            step_kwargs = {}
```
Comment (michaelosthege) [inline pymc3/sampling.py]: unused variable: `step_kwargs`

File: pymc3/step_methods/metropolis.py [hunk 4]
```diff
@@ -20,6 +21,9 @@ class Proposal:
     def __init__(self, s):
         self.s = s
 
+    def logp(self, s, value):
+        return multivariate_normal(np.zeros(s.shape[0]), cov=s).logpdf(value)
+
```
Comment (michaelosthege) [inline pymc3/step_methods/metropolis.py]: this logpdf does not apply to all `Proposal` distributions. Did you want to make it a method of the `MultivariateNormalProposal`?
Also consider adding a docstring describing the shapes required for `s` and `value`.

File: pymc3/step_methods/smc_ABC.py [hunk 5]
```diff
@@ -0,0 +1,418 @@
...
+    """
+    model = modelcontext(model)
+
+    if random_seed != -1:
+        np.random.seed(random_seed)
+
+    stage = 0
+    variables = model.vars
+    discrete = np.concatenate(
+        [[v.dtype in pm.discrete_types] * (v.dsize or 1) for v in variables])
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py]: Reformat for better readability:
```
discrete = np.concatenate([
    [v.dtype in pm.discrete_types] * (v.dsize or 1)
    for v in variables
])
```

File: pymc3/step_methods/smc_ABC.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+
+    return epsilon
+
+
+def calc_weights(un_weights):
+    """Compute importance weights"""
+    weights_un = np.exp(un_weights - un_weights.max())
+    weights = weights_un / np.sum(weights_un)
+    return weights
+
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py:314]: the `calc_weights` looks like a good candidate for a unit test

File: pymc3/step_methods/smc_ABC.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+        # resample based on plausibility weights (selection)
+        resampling_indexes = np.random.choice(
+            np.arange(draws), size=draws, p=weights)
+        posterior = posterior[resampling_indexes]
+        # compute proposal distribution based on weights
+        try:
+            covariance = _calc_covariance(posterior, weights)
+        except ValueError:
+            pm._log.info('Could not compute covariance matrix')
+            break
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py]: raise an informative error instead of failing silently

File: pymc3/step_methods/smc_ABC.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+        proposal = step.proposal(covariance)
+        # get distance function
+        distance_function = get_distance(step.distance_metric)
+
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py:170]: the comment in line 161 is trivial

File: pymc3/step_methods/smc_ABC.py [hunk 9]
```diff
@@ -0,0 +1,418 @@
...
+        # compute scaling and number of Markov chains steps (optional), based on previous
+        # acceptance rate
+        if step.tune and stage > 0:
+            if acc_rate == 0:
+                acc_rate = 1. / step.n_steps
+            step.scaling = _tune(acc_rate)
+            step.n_steps = 1 + \
+                (np.ceil(np.log(step.p_acc_rate) / np.log(1 - acc_rate)).astype(int))
+
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py:179]: the comment in 164/165 is good!
Would it make sense to move this computation into its own helper function? (And unit test it?)
Would `scaling` and `n_steps` be useful information for the `pm._log`?

File: pymc3/step_methods/smc_ABC.py [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+    if step.epsilons is None:
+        if step.routine == 'iqr':
+            range_iq = mquantiles(population, prob=[0.25, 0.75])
+            epsilon = np.abs(range_iq[1] - range_iq[0]) * iqr_scale
+        if step.routine == 'median':
+            epsilon = np.abs(np.median(population)) * iqr_scale
+    else:
+        epsilon = step.epsilons[stage]
+
+    return epsilon
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py:306]: also looks like a good unit test candidate. Remember to include corner-cases in the test case.

File: pymc3/step_methods/smc_ABC.py [hunk 11]
```diff
@@ -0,0 +1,418 @@
...
+    """
+    cov = np.cov(
+        np.squeeze(posterior_array),
+        aweights=weights.ravel(),
+        bias=False,
+        rowvar=0)
+    if np.isnan(cov).any() or np.isinf(cov).any():
+        raise ValueError(
+            'Sample covariances not valid! Likely "chains" is too small!')
+    return np.atleast_2d(cov)
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py:328]: Also a good test candidate. Here you definitely include a test case where the `ValueError` is raised.

File: pymc3/step_methods/smc_ABC.py [hunk 12]
```diff
@@ -0,0 +1,418 @@
...
+
+def get_distance(func_name):
+    d = {'absolute difference': absolute_difference,
+         'sum of squared_distance': sum_of_squared_distance,
+         'mean absolute_error': mean_absolute_error,
+         'mean squared_error': mean_squared_error,
+         'euclidean': euclidean_distance}
+    for key, value in d.items():
+        if func_name == key:
+            return value
```
Comment (michaelosthege) [inline pymc3/step_methods/smc_ABC.py]: the last three lines can be simplified to avoid a for loop:
```
if not func_name in d:
    raise KeyError('"{}" was not found in the standard distance metrics. Choose one of {}'.format(func_name, set(d.keys())))
return d[func_name]
```

**Review by michaelosthege — CHANGES_REQUESTED** [review]
Since I am not familiar with SMC/ABC myself, I can't really help with a review of the algorithm itself.

However, regarding performance issues you can use `%timeit` and related magic functions to find bottlenecks in a Jupyter notebook: https://jakevdp.github.io/PythonDataScienceHandbook/01.07-timing-and-profiling.html

**Comment by agustinaarroyuelo** [comment]
Thank you @michaelosthege for your review!

File: pymc3/distributions/simulator.py (hunk 2 above)
Comment (aloctavodia) [reply pymc3/distributions/simulator.py:50]: The random method of the simulator should be just the `function`, right?

File: pymc3/step_methods/smc_ABC.py (hunk 9 above)
Comment (agustinaarroyuelo) [reply pymc3/step_methods/smc_ABC.py:179]: Yes, It would make sense. This PR  #3373 is linked to that idea.

**Comment by agustinaarroyuelo** [comment]
On successive pull requests I will add the random method for the simulator and several tests suggested by @michaelosthege .

**Comment by aloctavodia** [comment]
Hi @michaelosthege, I think this should be ready to merge. You suggested 3 tests and Agustina did one. I suggest delaying the test on the epsilons as we have to revise how we are computing them and the other test you suggested I think it should be done directly on the common SMC/SMC-ABC functions #3373. Do you agree?

**Comment by michaelosthege** [comment]
Hi @aloctavodia and @agustinaarroyuelo ,
Sure, if you like to merge that's fine with me. (I'm on a conference and can't do another review right now.) Maybe you can then create issues for the things you want to address in a future PR?
cheers

**Review by twiecki — CHANGES_REQUESTED** [review]
Fix NB errors.

**Comment by twiecki** [comment]
I think the NB would really benefit from a bit more exposition on ABC (can be high level) and more detail on the code/API. The way it looks to me it's the only piece of documentation we would currently have.

**Comment by agustinaarroyuelo** [comment]
It is a good idea

**Comment by twiecki** [comment]
@agustinaarroyuelo Thanks, that reads quite well. Typos:
* "ABC comes" -> "ABC becomes"
* iteratevly -> iteratively

Also, what is up with the traceplot? I understand the chain is probably nonsensical to look at but it doesn't look like parameters are changed at all and are fixed in two modes. These somehow seem to average to something close to the true value (why?). The resulting posterior predictive plot also don't look super convincing and would be better too if it didn't use the mean but make use of the fact that this is in a Bayesian framework.

**Comment by aloctavodia** [comment]
@twiecki the current implementation is not usable yet at it seems very sensitive to the values of the parameters like the epsilons, the problems you mention are part of what we need to address. I would like to have a usable version (although maybe still experimental) for the next release.

**Comment by twiecki** [comment]
@aloctavodia OK, what should we do with this PR then? Park it?

**Comment by aloctavodia** [comment]
I will like to accept it and keep working on SMC-ABC in subsequent PRs: I will be happy to work on this with @agustinaarroyuelo. But if you think is too unstable for merging then we can close it and wait for a more stable version.

**Comment by junpenglao** [comment]
Fine to merge first - but could you please add some comments about this is not currently working?

**Comment by twiecki** [comment]
I disagree, I don't think we should merge things we know not to work, that
would just confuse users (and us). Rather we should keep it on a branch
until it's ready.

**Comment by junpenglao** [comment]
Is there any other (toy) model that demonstrate it works in principle?

**Comment by aloctavodia** [comment]
@agustinaarroyuelo had a very simple Gaussian model I think. But I agree with @twiecki, having this on master could be confusing.

**Comment by twiecki** [comment]
If you want we could merge this into https://github.com/pymc-devs/pymc3-experimental.

**Comment by agustinaarroyuelo** [comment]
After several discussions with @aloctavodia  we decided to refactor both the SMC and SMC-ABC code. The idea is that both step methods should share the most components as possible. So, I am closing this pull request, and @aloctavodia is going to open a draft PR, to continue the discussion on this project.
